File: project/jooeun-onboarding/survey-common/src/main/java/com/innercircle/survey/common/util/UlidGenerator.java
package com.innercircle.survey.common.util;

import de.huxhorn.sulky.ulid.ULID;

/**
 * ULID (Universally Unique Lexicographically Sortable Identifier) 생성 유틸리티
 * 
 * 분산 환경에서 고유하고 시간순 정렬이 가능한 ID를 생성합니다.
 * - UUID 대비 시간 순서 보장
 * - 데이터베이스 인덱스 성능 향상
 * - 다중 인스턴스에서 충돌 없는 ID 생성
 */
public final class UlidGenerator {

    private static final ULID ulid = new ULID();

    private UlidGenerator() {
        throw new AssertionError("인스턴스화할 수 없습니다.");
    }

    /**
     * 새로운 ULID 문자열을 생성합니다.
     *
     * @return 26자리 ULID 문자열 (예: 01ARZ3NDEKTSV4RRFFQ69G5FAV)
     */
    public static String generate() {
        return ulid.nextULID();
    }

    /**
     * 지정된 타임스탬프로 ULID를 생성합니다.
     *
     * @param timestamp 밀리초 단위 타임스탬프
     * @return ULID 문자열
     */
    public static String generate(long timestamp) {
        return ulid.nextULID(timestamp);
    }

    /**
     * ULID에서 타임스탬프를 추출합니다.
     *
     * @param ulidString ULID 문자열
     * @return 밀리초 단위 타임스탬프
     */
    public static long extractTimestamp(String ulidString) {
        return ULID.parseULID(ulidString).timestamp();
    }
}

File: project/sanghyuk-onboarding/src/main/kotlin/fc/innercircle/sanghyukonboarding/common/infrastructure/numbering/UlidGenerator.kt
package fc.innercircle.sanghyukonboarding.common.infrastructure.numbering

import java.security.SecureRandom

/**
 * ULID (Universally Unique Lexicographically Sortable Identifier) Generator
 *  - 48-bit timestamp (milliseconds since unix epoch)
 *  - 80-bit randomness
 *  - Crockford's Base32 encoding → 26-character string
 */
object UlidGenerator {
    private val ENCODING = charArrayOf(
        '0','1','2','3','4','5','6','7','8','9',
        'A','B','C','D','E','F','G','H','J','K',
        'M','N','P','Q','R','S','T','V','W','X','Y','Z'
    )

    private val random = SecureRandom()

    /**
     * Generates a new ULID string (26 characters)
     */
    fun next(): String {
        // 1) Timestamp (48 bits)
        val time = System.currentTimeMillis()
        val timeChars = CharArray(10)
        var value = time
        for (i in 9 downTo 0) {
            timeChars[i] = ENCODING[(value and 0x1FL).toInt()]
            value = value ushr 5
        }

        // 2) Randomness (80 bits)
        val randomBytes = ByteArray(10)
        random.nextBytes(randomBytes)
        val randChars = CharArray(16)
        var bitBuffer = 0L
        var bitBufferLen = 0
        var byteIndex = 0
        var charPos = 0
        while (charPos < 16) {
            if (bitBufferLen < 5) {
                bitBuffer = (bitBuffer shl 8) or (randomBytes[byteIndex].toLong() and 0xFFL)
                bitBufferLen += 8
                byteIndex++
            }
            bitBufferLen -= 5
            val idx = ((bitBuffer ushr bitBufferLen) and 0x1F).toInt()
            randChars[charPos++] = ENCODING[idx]
        }

        return String(timeChars) + String(randChars)
    }
}
